### PR TITLE
Teach TypeReconstruction how to look through associated types of the …

### PR DESCRIPTION
**Pull request by egranata**
Code to make TypeReconstruction able to discover associated types of Self protocols, e.g. Iterator in

extension Collection { func foo(_ x: Iterator) {...} }

…Self protocol

### CODE REVIEW

**Comment by egranata** [comment]
@swift-ci please smoke test

**Comment by egranata** [comment]
Failed in llbuild:
"<unknown>:0: error: missing input 'input' and no rule to build it
/bin/sh -c true
<unknown>:0: error: build had 1 command failures

note: command had no output on stdout or stderr
error: command failed with exit status: 1"

I don't think this is related to my change...

**Comment by egranata** [comment]
@swift-ci please smoke test Linux platform

File: lib/IDE/TypeReconstruction.cpp [hunk 1]
```diff
@@ -866,14 +866,27 @@ static void VisitNodeAssociatedTypeRef(
...
-        if (archetype->hasNestedType(identifier)) {
-          Type nested = archetype->getNestedTypeValue(identifier);
-          if (nested) {
-            result._types.push_back(nested);
-            result._module = type_result._module;
-            return;
+        Type nested;
+        if (archetype->hasNestedType(identifier))
+          nested = archetype->getNestedTypeValue(identifier);
+        else if (ProtocolDecl *self_protocol = archetype->getSelfProtocol()) {
```
Comment (DougGregor) [inline lib/IDE/TypeReconstruction.cpp]: @slavapestov, aren't we trying to phase out uses of `getSelfProtocol()`?

Comment (egranata) [inline lib/IDE/TypeReconstruction.cpp]: If so, I am going to need a different way to do the same thing

Comment (slavapestov) [inline lib/IDE/TypeReconstruction.cpp]: The `archetype->hasNestedType(identifier)` check should always succeed. Why are we hitting the second case?

And yes, `getSelfProtocol()` is deprecated. Please do not introduce new usages.

Comment (slavapestov) [inline lib/IDE/TypeReconstruction.cpp]: Discussed in person. What Enrico wants to do still requires this hack, but a patch I have in another PR cleans this up (basically by nuking the SelfTypeRef mangling altogether). Go ahead and merge this PR if you want.

**Comment by slavapestov** [comment]
Looks good for CCC.

**Pull request merged**